Various improvements to the tutorial and building experience.

housejester opened the pull request:
- Attempt to detect and set java_home if not set. When building locally, I have java 6 and 7 on my Mac. Even though 7 is the 'default', mvn still picks up 6 and the build fails. This just does a best effort to set java_home in the script and detect it supports the required source version.
- Clarify docs on building & where to get artifact. When building from source, it wasn't obvious what to do /after/ building from source (ie grab the services tarball).
- Fix druid VERSION resolution in build script. Seems the latest sonatype work broke the VERSION lookup in the buildscript (was finding '2' which is the parent pom version).
- Improve the run_example_server and client scripts. I made them be such that the user selects by a number and it will list all the files matching the pattern (for example *query.body for the client). This makes it so that other examples can be more easily dropped in and used. It still supports the original usage of calling the script with an example 'name' arg.
- Retry geo ip download if corrupt. If you kill the realtime node at the wrong time (ie during startup), the geo ip file is only partially downloaded and corrupted. This leads to frustration b/c it is failing and not indicating where the file is. Further, on the Mac, the file isn't landing in /tmp, but a /var/folders/... tmp space, so it was difficult to find so I could delete it. So this just retries the download as well as streaming into a .tmp extension until fully downloaded. It won't do any re-downloading if the path is specified in the config. I could split this one out if desirable.

drcrallen commented on server/src/main/java/io/druid/segment/realtime/firehose/WikipediaIrcDecoder.java:
```diff
@@ -90,35 +90,59 @@ public WikipediaIrcDecoder(
-    } else {
+  private DatabaseReader openGeoDB(String configDbPath) {
+    RuntimeException lastEx = null;
+    for (int i=0; i<2; i++) {
```
Why is this running twice?

housejester replied:
It is a retry for when the config does not have a geo db file path specified (ie, we're downloading it into a temp space), but opening that file results in an exception. If it opens fine, it doesn't retry. If it throws an exception, but it was specified in the config, it will also not retry. 

In my case, the decoder was failing in the constructor because the file was only partially downloaded (process was killed mid-download), so the server was never coming up. The thought was that if it could detect this and just try to re-download the db, it could recover without user intervention (where the user has to find and delete the file). So, this retry will delete it and then attempt to download and open it again. However, the retry may be not be necessary with the change that has the download writing to a .tmp extension until fully downloaded.

housejester replied:
I updated this to be (hopefully) more clear.

fjy commented:
@housejester Thanks very much for this contrib! We will code review it. Is it cool if we get this in for 0.7.1?

fjy commented on server/src/main/java/io/druid/segment/realtime/firehose/WikipediaIrcDecoder.java:
```diff
@@ -90,35 +90,59 @@ public WikipediaIrcDecoder(
+    File geoDb;
     try {
-      geoLookup = new DatabaseReader(geoDb);
+      String tmpDir = System.getProperty("java.io.tmpdir");
```
I actually ended up implementing a very similar fix, which is probably generating the merge conflict. I think we submitted our PRs a few mins apart from each other :D

housejester replied:
Cool. I'll rebase this and merge in yours.

fjy commented on docs/content/Tutorial:-A-First-Look-at-Druid.md:
````diff
@@ -72,13 +78,13 @@ You should see a bunch of files:
 ```
 
-Select "wikipedia".
+Select "2" for the "wikipedia" example.
````
is it possible to support both selecting a number and typing it in?

The reason I ask is I always run with ./run_example_server.sh wikipedia

housejester replied:
That form still works :) actually, when specified as an arg to the script like that, it will look for examples/{name}/{name}_runtime.properties the same as it did before. if that file isn't found it falls back to prompting again.

housejester commented:
0.7.1 is fine by me.

drcrallen commented on server/src/main/java/io/druid/segment/realtime/firehose/WikipediaIrcDecoder.java:
```diff
@@ -90,35 +90,59 @@ public WikipediaIrcDecoder(
-      geoLookup = new DatabaseReader(geoDb);
+      String tmpDir = System.getProperty("java.io.tmpdir");
+      geoDb = new File(tmpDir, this.getClass().getCanonicalName() + ".GeoLite2-City.mmdb");
+      if (!geoDb.exists()) {
```
If it does exist, can you print an info message that you've found the downloaded file and the location it was found? That way the user is aware we have files residing on their system.

housejester replied:
The full path is now logged in a few places: just before downloading, on failed open, and on successful open. Does that work?

drcrallen commented:
This shouldn't block the PR but is a good question to raise here:

Is there a good way to unit test the examples?

housejester commented:
I need to update this since part of it is now made irrelevant by https://github.com/druid-io/druid/pull/1145

fjy commented on build.sh:
```diff
@@ -2,11 +2,30 @@
 BASE=$(cd $(dirname $0) && pwd)
 
-VERSION=`cat pom.xml | grep '<version>' | head -1 | tail -1 | sed 's_.*<version>\([^<]*\)</version>.*_\1_'`
+VERSION=`cat pom.xml | grep -A1 '<artifactId>druid</artifactId>' | grep '<version>' | head -1 | sed 's_.*<version>\([^<]*\)</version>.*_\1_'`
```
@housejester,  @xvrl actually removed the build script because it was unnecessary and not very useful. It is easier to just use mvn package.

fjy commented:
I can merge this after resolving the conflict in build.sh

sharrissf commented:
@drcrallen For samples it's not unit testing but having automated tests for them is as critical or more critcal than the rest of the system as it impacts adoption. I would check in with Gautam on the best way to approach it.

fjy commented:
@drcrallen Automated testing for examples is a larger discussion and I don't think we need to have it here. But I agree it is very important and should be proposed to the community.

@sharrissf Please use github tags for individuals so the rest of the community knows who you are referring to.

housejester commented:
I removed build.sh, and rebased. Now getting

```bash
2015-02-26T02:41:55,275 INFO [main] io.druid.initialization.Initialization - Loading extension[io.druid.extensions:druid-examples] for class[io.druid.cli.CliCommandCreator]
2015-02-26T02:41:56,647 ERROR [main] io.druid.initialization.Initialization - Unable to resolve artifacts for [io.druid.extensions:druid-examples:jar:0.7.1-SNAPSHOT (runtime) -> [] < [central (http://repo1.maven.org/maven2/, releases+snapshots),  (https://metamx.artifactoryonline.com/metamx/pub-libs-releases-local, releases+snapshots)]].
java.lang.NullPointerException
```
when trying to run-example-server.sh. I'm assuming it is trying to pull the snapshot, but it isn't up yet.

xvrl commented:
@housejester we don't publish snapshot artifacts for Druid in maven, the examples are meant to be run off of the release tarball. If you want to try locally with snapshots, make sure to run `maven install` first, that should fix it.

housejester commented:
Yeah. I guess the surprising part was that I built locally with mvn package (per the 'building from source' docs), but then at runtime, it tries to fetch snapshots again. Which I understand is from the module system (I've not looked into that part much yet). I think this should either be mentioned in the building from source docs, AND/OR detected and logged as a hint (vs a null pointer exception) when this happens (ie snapshots requested and not found...).

However, `mvn install` did indeed fix the issue for me :)
